src/qat/purr/backends/qblox/codegen.py

```diff
@@ -374,36 +396,119 @@ def delay(self, inst: Delay):
         self._duration = self._duration + inst.duration
 
     def waveform(self, waveform: Waveform, target: PulseChannel):
-        i_index, q_index, num_samples = self._process_waveform(waveform, target)
-        if (i_index, q_index, num_samples) == (None, None, 0):
+        pulse = self._evaluate_waveform(waveform, target)
+        num_samples = pulse.size
+
+        if num_samples < Constants.GRID_TIME:
+            if num_samples == 0:
+                log.warning(f"Empty pulse.")
+            else:
+                log.warning(
+                    f"""
+                    Minimum pulse width is {Constants.GRID_TIME} ns with a resolution of {1} ns.
+                    Please round up the width to at least {Constants.GRID_TIME} nanoseconds.
+                    """
+                )
+            log.warning("This pulse will be ignored.")
             return
-        self.sequence_builder.play(i_index, q_index, Constants.GRID_TIME)
-        self.sequence_builder.wait(num_samples - Constants.GRID_TIME)
+
+        if isinstance(waveform, Pulse) and waveform.shape == PulseShapeType.SQUARE:
+            i_offs_steps = int(
+                pulse[0].real * (Constants.MAX_OFFSET_SIZE // 2)  # Signed integer
+            )
+            q_offs_steps = int(
+                pulse[0].imag * (Constants.MAX_OFFSET_SIZE // 2)  # Signed integer
+            )
+            self.sequence_builder.set_awg_offs(i_offs_steps, q_offs_steps)
+            self.sequence_builder.upd_param(num_samples)
+            self.sequence_builder.set_awg_offs(0, 0)
+            self.sequence_builder.upd_param(Constants.GRID_TIME)
+        else:
+            if 2 * num_samples > self._wf_memory:  # for both I and Q
+                raise ValueError(
+                    f"No more waveform memory left for {waveform} on pulse channel {target}"
+                )
+
+            wf_name = type(waveform).__name__
+            if isinstance(waveform, Pulse):
+                wf_name = waveform.shape.name
+
+            i_index, q_index = self._register_waveform(wf_name, pulse)
+            self.sequence_builder.play(i_index, q_index, num_samples)
+
         self._duration = self._duration + waveform.duration
 
     def measure_acquire(
```

Comment:
this is a very large method with nested functions. Please refactor to be cleaner and remove nested functions. This gives very unclear behaviour and makes granular testing very difficult